### DISCUSSION TITLE
HLS on iOS

### DISCUSSION THREAD

**Discussion by nadecancode** [shiyiya/oplayer#23] [General]
#21

**Comment by shiyiya** [comment]
url

**Reply by nadecancode** [reply]
https://enime.moe/watch/call-of-the-night/10

Wait I'm going to deploy the version that has 1.0.33

EDIT: Deployed

**Comment by nadecancode** [comment]
Issue still persists in 1.0.33, source loaded with ``changeSource`` can be correctly played and only initial source cannot be loaded natively on iOS

**Comment by nadecancode** [comment]
Example implementation here https://github.com/Enime-Project/enime.moe/blob/8164e44b361b8e1eeb6e5ce77bed8c877fbbad0b/components/player/index.vue#L36

**Comment by shiyiya** [comment]
![image](https://user-images.githubusercontent.com/34017352/189285529-2887f66f-2588-4b31-bc17-db067a921595.png)

try refresh cache.

**Reply by nadecancode** [reply]
Opened in incognito which is not supposed to have cache, still same result.

**Comment by shiyiya** [comment]
![image](https://user-images.githubusercontent.com/34017352/189286720-245a67a5-d260-421b-8ca0-d3ceaee58b33.png)

https://cache.387e6278d8e06083d813358762e0ac63.com/222870623924.m3u8

**Comment by nadecancode** [comment]
Ok just tried the implementation in example and it works. Weird.

**Comment by shiyiya** [comment]
![image](https://user-images.githubusercontent.com/34017352/189288806-1a36b9e3-77d2-48d5-a532-8415bfdb0867.png)

💯

**Reply by nadecancode** [reply]
The heck, it doesn't play on my actual iPhone lmao?

**Reply by shiyiya** [reply]
Maybe there is a difference between the emulator and the real machine?

**Reply by nadecancode** [reply]
the real machine is able to play the example in repository but not on the site

**Reply by shiyiya** [reply]
可能你链接404了 Try more

**Reply by nadecancode** [reply]
Tried a lot already lol, it is able to play on standalone example with same link

**Comment by shiyiya** [comment]
Works well on the emulator

**Comment by nadecancode** [comment]
Weirdly, if loaded by changeSource it successfully plays

**Comment by nadecancode** [comment]
Ok I'm suspecting it's because of SSR, gimme few moments

**Comment by nadecancode** [comment]
Nevermind still no luck, it's so weird

**Comment by nadecancode** [comment]
Ok nevermind, I don't know why but the site is only able to play backup source properly instead of triggering with changeSource, this is even weirder because they're both m3u8 and only iOS cannot play this @shiyiya

**Comment by nadecancode** [comment]
@shiyiya LMAO what the heck, the subtitle option in UI plugin is the gamechanger

**Reply by nadecancode** [reply]
If I remove subtitle option the primary source will be able to play

**Comment by nadecancode** [comment]
@shiyiya OK I think it's because of how iOS handles subtitles natively. Currently I have this ```ts
ui({
        fullscreenWeb: false,
        subtitle: {
          source: sourceRef.value.subtitle ? [
            {
              default: true,
              src: sourceRef.value.subtitle
            }
          ] : [],
          fontSize: 30,
          enabled: true
        }
      })])
      ``` since subtitlechange event only accepts source payload, and since primary source has burnt-in subtitles (means no external ones, leading to empty subtitle array) the iOS will fail to render to video because OPlayer tells iOS devices that subtitles are there but in reality subtitle is an empty array (since enabled is true)

**Reply by nadecancode** [reply]
There are 2 solutions that I can think of:
1. Make subtitlechange accept something more than source payload so I can set "enabled" property properly
2. Don't append subtitle element (track) if subtitle array is empty

**Reply by shiyiya** [reply]
https://github.com/shiyiya/oplayer/blob/main/packages/ui/src/components/Subtitle.ts#L172

In fact, if the array is empty, he won't do anything

**Reply by nadecancode** [reply]
> https://github.com/shiyiya/oplayer/blob/main/packages/ui/src/components/Subtitle.ts#L172
> 
> In fact, if the array is empty, he won't do anything

https://github.com/shiyiya/oplayer/blob/96a2f960e1dd5dcb106580fd1339604267c52e12/packages/ui/src/components/Subtitle.ts#L75
I think it's more likely related to this

**Reply by nadecancode** [reply]
The track element is getting created no matter what

**Reply by nadecancode** [reply]
https://github.com/shiyiya/oplayer/blob/96a2f960e1dd5dcb106580fd1339604267c52e12/packages/ui/src/components/Subtitle.ts#L138

If currentSubtitle does not exist, then the track src will not be set but the track element is there. I think this is the different between iOS real device and emulator that the real device cannot accept a track element with no src

**Reply by nadecancode** [reply]
> https://github.com/shiyiya/oplayer/blob/96a2f960e1dd5dcb106580fd1339604267c52e12/packages/ui/src/components/Subtitle.ts#L138
> 
> If currentSubtitle does not exist, then the track src will not be set but the track element is there. I think this is the different between iOS real device and emulator that the real device cannot accept a track element with no src

Or it's chrome behavior, since emulator uses Safari

**Reply by shiyiya** [reply]
可以测试一下ios真机能不能空track， 这里说了 只有currentSubtitle不为空才会被调用

**Reply by nadecancode** [reply]
应该是Chrome的问题, Safari没这问题

**Comment by shiyiya** [comment]
![HC_20220909_210342_null](https://user-images.githubusercontent.com/34017352/189356294-f8097360-5344-49d4-a126-f38b23fac4a9.jpg)

**Reply by shiyiya** [reply]
can control the loading and progress but clicking play video will not start

**Comment by nadecancode** [comment]
Tested latest patch and it works 💯
